keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +133,75 @@ A typical successful issuance proceeds as follows.
    the `.Status.Certificate` field.
 1. The requestor observes the update, and stores the certificate locally.
 
+### Signers
+CSRs have a `signerName` field which is used to specify which signer the CSR creator wants to sign the certificate.
+To support migration from v1beta1 to v1, this required field will be defaulted in v1beta1 (optional in openapi), but
+not defaulted and required in v1 :
+ 1. If it's a kubelet client certificate, it is assigned "kubernetes.io/kubelet-client".
+ 2. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+
+There will be field selector support to make approvers and signers easier to write.
+
+All signers should provide information about how they work so that clients can predict what will happen to their CSRs.
+This includes:
+ 1. Trust distribution - how trust (ca bundles) are distributed.
+ 2. Permitted subjects - (any? specific subtree?) and behavior when a disallowed subject is requested.
+ 3. Permitted x509 extensions - (IP SANs? DNS SANs? Email SANs? URI SANs? others?) and behavior when a disallowed
+ extension is requested.
+ 4. Permitted key usages / extended key usages - (client only? server only? any? signer-determined? CSR-determined?) and
+ behavior when usages different than the signer-determined usages are specified in the CSR.
+ 5. Expiration/cert lifetime - (fixed by signer? configurable by admin? CSR-determined?) and behavior when an expiration
+ different than the signer-determined expiration is specified in the CSR.
+ 6. CA bit allowed/disallowed - and behavior if a CSR contains a request a for a CA cert when the signer does not permit it.
+
+
+sig-auth reserves all `kubernetes.io/*` `signerNames` and more may be added in the future.
+Kubernetes provides the following well-known signers.  Today, failures for all of these are only reported in kube-controller-manager logs:
+ 1. kubernetes.io/kube-apiserver-client - signs certificates that will be honored as client-certs by the kube-apiserver.
+    Never auto-approved by kube-controller-manager.
```

Comment:
Never auto-approved means this signer's CSR only will be approved by human.

How dose a people(cluster admin) judge if this CSR should be approved only according to the fields in the CSR? Should he copy the `CSR.Request` and decode it in the his personal computer to see what's in it?

I do not think `kubernetes.io/kube-apiserver-client` and `kubernetes.io/legacy-unknown` is meaningful. On the other words, if Kubrentes can not auto-approve the signer's CSR, then the signer is not meaningful, and should be removed.

Reply:
> Never auto-approved means this signer's CSR only will be approved by human.

It means the kube-controller-manager will not auto-approve. Specific deployments can set up approver controllers that make judgements based on information specific to the cluster.

> if Kubrentes can not auto-approve the signer's CSR, then the signer is not meaningful, and should be removed.

I disagree. The existing kubelet serving certificate signer is a good example of a standard certificate that can be requested, but which requires approval using cluster-specific knowledge. Decoupling the request and signer from the approver is useful.

Reply:
> Specific deployments can set up approver controllers that make judgements based on information specific to the cluster.

I do not think this approver will be exist. Because the approver can not get the identity who create this CSR. Which means a NOBODY can request a CSR with CommonName `cluster-admin`, and the approver can not verify.

Two month ago, I want to provide a solution to self-sign `kubeconfig` file for my company developers, based on [cert-manager](https://github.com/jetstack/cert-manager). User can create `Certificate`(a CRD of `cert-manager`, just like a CSR) , then he can get the certificate/key and kubeconfig. But this solution is total failed, as the we have no way to verify the CommonName in `Certificate` and the real user who create this `Certificate`.

Reply:
> I do not think this approver will be exist. Because the approver can not get the identity who create this CSR. Which means a NOBODY can request a CSR with CommonName `cluster-admin`, and the approver can not verify.

The identity that created the CSR is recorded in the CSR spec. The same information used by the kube-controller-manager to auto-approve is available to any other controller via the CSR API.

Reply:
> The identity that created the CSR is recorded in the CSR spec (in the username/uid/groups/extra fields)

I see, that's the difference from my `cert-manager` solution. Thanks for your explain.